app/publish/publish.go

```diff
@@ -1,131 +1,127 @@
-	// var rpcParams *ljsonrpc.StreamCreateOptions
-	rpcParams := struct {
-		Name                          string  `json:"name"`
-		Bid                           string  `json:"bid"`
-		FilePath                      string  `json:"file_path,omitempty"`
-		FileSize                      *string `json:"file_size,omitempty"`
-		IncludeProtoBuf               bool    `json:"include_protobuf"`
-		Blocking                      bool    `json:"blocking"`
-		*ljsonrpc.StreamCreateOptions `json:",flatten"`
-	}{}
-
-	query, err := proxy.NewQuery(rawQuery)
-	if err != nil {
-		panic(err)
-	}
-
-	if err := query.ParamsToStruct(&rpcParams); err != nil {
-		panic(err)
-	}
+// Publish takes a file path, account ID and client JSON-RPC query,
+// patches the query and sends it to the SDK for processing.
+// Resulting response is then returned back as a slice of bytes.
+func (p *LbrynetPublisher) Publish(filePath, accountID string, rawQuery []byte) []byte {
+	c := p.Service.NewCaller()
+	c.SetAccountID(accountID)
+	c.SetPreprocessor(func(q *proxy.Query) {
+		params := q.ParamsAsMap()
+		params[fileNameParam] = filePath
+		q.Request.Params = params
+	})
+	r := c.Call(rawQuery)
+	return r
+}
 
-	if rpcParams.FilePath != "__POST_FILE__" {
-		panic("unknown file_path content")
+func (h UploadHandler) handleUpload(r *users.AuthenticatedRequest) (*os.File, error) {
+	file, header, err := r.FormFile("file")
+	if err != nil {
+		return nil, err
 	}
+	defer file.Close()
 
-	bid, err := strconv.ParseFloat(rpcParams.Bid, 64)
-	rpcParams.FilePath = filePath
-	rpcParams.AccountID = &accountID
-
-	result, err := lbrynet.Client.StreamCreate(rpcParams.Name, filePath, bid, *rpcParams.StreamCreateOptions)
+	f, err := h.CreateFile(r.AccountID, header.Filename)
 	if err != nil {
 		return nil, err
 	}
 
-	rpcResponse := jsonrpc.RPCResponse{Result: result}
-	serialized, err := json.MarshalIndent(rpcResponse, "", "  ")
-	if err != nil {
+	if _, err := io.Copy(f, file); err != nil {
+		return nil, err
+	}
+	if err := f.Close(); err != nil {
 		return nil, err
 	}
-	return serialized, nil
+	return f, nil
 }
 
-func (h uploadHandler) Handle(w http.ResponseWriter, r *users.AuthenticatedRequest) {
+// Handle is where HTTP upload is handled and passed on to Publisher.
+// It should be wrapped with users.Authenticator.Wrap before it can be used
+// in a mux.Router.
+func (h UploadHandler) Handle(w http.ResponseWriter, r *users.AuthenticatedRequest) {
+	w.WriteHeader(http.StatusOK)
+
 	if !r.IsAuthenticated() {
 		var authErr Error
 		if r.AuthFailed() {
 			authErr = NewAuthError(r.AuthError)
 		} else {
 			authErr = ErrUnauthorized
 		}
-		w.WriteHeader(http.StatusOK)
 		w.Write(authErr.AsBytes())
 		return
 	}
-	file, header, err := r.FormFile("file")
-	if err != nil {
-		panic(err)
-	}
-	defer file.Close()
 
-	f, err := h.CreateFile(r.AccountID, header.Filename)
+	f, err := h.handleUpload(r)
 	if err != nil {
-		panic(err)
+		w.Write(NewInternalError(err).AsBytes())
```

Comment:
This is an example of where I feel like we are not using a pattern that we have already defined with internal-apis. I see where you are adding code, now in multiple PRs that I have reviewed. I feel like the internal-apis pattern is a much better design and makes adding apis super easy. This is a lot of manual code handling. take a look at the api server package in lbry.go. I know you did not start from this ( not sure why grin didn't have you start from there ), but maybe you can leverage some of the code there to make your work easier. I see you writing to the responsewriters all the time. In the apiserver package we have a generic function that handles this for us so we never have to write a line like you do above. We have a special type for response and we can leverage the status automatically like `api.Response{Error:err, Status: http.StatusBadRequest}` and we can use things like `api.StatusError{Err:err,Status: http.StatusNotAuthenticated}` which returns an error and the code knows how to handle that. All regular errors are automatically server errors, and the stack traces get sent along with he response automatically etc. It just seems like a much better pattern. Not sure if you are aware of it. I think you would like it alot better.

Reply:
Grin put me on that path with his initial commit so I'm aware of that pattern. Unfortunately I had to abandon it straight away because one important distinction of my API endpoints is that they seldom generate the responses they're sending to the client. It's mostly errors, and they're not terribly complicated. Even in this package I'm simply passing down the wire the `[]byte` response I get from the SDK without any attempt to parse it. Hence I cannot have my views simply return `api.Response` structs.

I may be missing something but it seems for me to use `lbry.go/server` machinery here would be more trouble than it's worth.